example/gpu_memory_profiler/README.md

```diff
@@ -0,0 +1,151 @@
+
+![MXNet-GPU_Memory_Profiler-Design](./MXNet-GPU_Memory_Profiler-Design.png)
+
+
+## Instructions
+
+(*The video below is shown at **SysML 2019** ([Demo# 24](https://www.sysml.cc/doc/2019/demo_24.pdf))*.)
+
+![MXNet-GPU_Memory_Profiler](./sysml19_demo/MXNet-GPU_Memory_Profiler.gif)
+
+*In order for the GPU memory profiler to be enabled, you need to **compile from source**.*
+
+- Download the MXNet codebase and install the prerequisite software libraries 
+    (e.g., *OpenCV*, *jemalloc*, etc.).
+
+```bash
+git clone --recursive https://github.com/apache/incubator-mxnet mxnet
+sudo apt-get install ...
+```
+
+- Modify the `MXNET_ENABLE_STORAGE_TAGGING` flag in `include/storage_tag.h` to **1**,
+    which controls the storage tagging (disabled by default),
+    **and** `MXNET_USE_GPU_MEMORY_PROFILER` flag in `src/profiler/gpu_memory_profiler.h` **1**,
+    which controls dumping the GPU memory allocation inforamtion (also disabled by default).
+  Please note that you must have **BOTH flags** set to **1** to use the GPU memory profiler.
+
+- Build the MXNet core library.
+
+```bash
+cd mxnet # workspace/mxnet
+make -j $(nproc) USE_OPENCV=1 \
+                 USE_BLAS=openblas \
+                 USE_CUDA=1 \
+                 USE_CUDA_PATH=/usr/local/cuda \
+                 USE_CUDNN=1
+```
+
+- Build the Python binding.
+
+```bash
+cd python # workspace/mxnet/python
+python3 setup.py install
+```
+
+- Run your application, as normal, but with `MXNET_USE_GPU_MEMORY_PROFILER`
+    environment variable set to **1**.
+  You should be able to see MXNet telling you that "*MXNet has memory profiler enabled.*".
+
+```bash
+env MXNET_USE_GPU_MEMORY_PROFILER=1 MXNET_ENGINE_TYPE=NaiveEngine python3 ... # your application goes here
+```
+
+- You shall be able to see two files created in your current working directory,
+    namely `mxnet_gpu_memory_profiler_output.csv` and 
+           `mxnet_gpu_memory_profiler_output.log`.
+  The former stores the mapping between the storage tag and the allocation size (in `MiB`).
+  The latter tracks the call stacks that lead to an unknown storage tag (for debugging purpose).
+
+- Run the `two_bar_plot.py` and `pprint_top_entries.py` for plotting and analyzing 
+    information on the profiled information.
+  Please note that you need to customize the keyword dictionary and the path to 
+    the memory profile `mxnet_gpu_memory_profiler_output.csv` in `SETME.py` 
+    **first** before running the analyzing scripts.
+
+
+## !Attention
+
+*As you use the GPU memory profiler, please kindly pay close attention to the followings.*
+
+- The information reported by the GPU memory profiler depends on the symbol names
+    you assign at the Python frontend.
+  Therefore, it is important that you give meaningful names as you program your models.
+- It is **STRONGLY RECOMMENDED** that you use the **NAIVE** engine and storage pool for the GPU memory profiling
```

Comment:
Should we set it ourselves inside the engine? GPU_PROFILE => NAIVE ENGINE

Reply:
The GPU memory profiler also works for non-naive engine types. The problem with non-naive engines is that they can make the GPU memory profiling logs out-of-order and hence hard to read and digest. This is why I added it as a suggestion rather than a strong restriction.

Reply:
Thanks for the clarification